src/components/Trade/TradeInputV2.tsx

```diff
@@ -28,54 +39,91 @@ import { type TradeState, TradeAmountInputField, TradeRoutePaths } from './types
 const moduleLogger = logger.child({ namespace: ['TradeInput'] })
 
 export const TradeInput = () => {
-  useSwapperService()
+  const { isLoadingTradeQuote, isLoadingFiatRateData } = useSwapperService()
   const [isLoading, setIsLoading] = useState(false)
-  const { setTradeAmounts } = useTradeAmounts()
-  const { checkApprovalNeeded, getTrade } = useSwapper()
+  const { setTradeAmountsAsynchronous, setTradeAmountsSynchronous } = useTradeAmounts()
+  const { checkApprovalNeeded, getTrade, bestTradeSwapper } = useSwapper()
   const history = useHistory()
   const borderColor = useColorModeValue('gray.100', 'gray.750')
   const { control, setValue, getValues, handleSubmit } = useFormContext<TradeState<KnownChainIds>>()
+  const {
+    state: { wallet },
+  } = useWallet()
 
+  // Watched form fields
   const sellTradeAsset = useWatch({ control, name: 'sellTradeAsset' })
   const buyTradeAsset = useWatch({ control, name: 'buyTradeAsset' })
   const quote = useWatch({ control, name: 'quote' })
   const feeAssetFiatRate = useWatch({ control, name: 'feeAssetFiatRate' })
+  const buyAssetFiatRate = useWatch({ control, name: 'buyAssetFiatRate' })
   const fees = useWatch({ control, name: 'fees' })
   const sellAssetAccountId = useWatch({ control, name: 'sellAssetAccountId' })
   const buyAssetAccountId = useWatch({ control, name: 'buyAssetAccountId' })
   const fiatSellAmount = useWatch({ control, name: 'fiatSellAmount' })
   const fiatBuyAmount = useWatch({ control, name: 'fiatBuyAmount' })
   const slippage = useWatch({ control, name: 'slippage' })
 
-  const translate = useTranslate()
-
+  // Selectors
   const sellFeeAsset = useAppSelector(state =>
     selectFeeAssetById(state, sellTradeAsset?.asset?.assetId ?? ethAssetId),
   )
-  const sellAssetBalance = useAppSelector(state =>
+  const feeAssetBalance = useAppSelector(state =>
+    sellFeeAsset
+      ? selectPortfolioCryptoHumanBalanceByAssetId(state, { assetId: sellFeeAsset?.assetId })
+      : null,
+  )
+  const sellAssetBalanceCrypto = useAppSelector(state =>
     selectPortfolioCryptoBalanceByFilter(state, {
       accountId: sellAssetAccountId,
       assetId: sellTradeAsset?.asset?.assetId ?? '',
     }),
   )
 
-  const protocolFee = fromBaseUnit(bnOrZero(fees?.tradeFee), buyTradeAsset?.asset?.precision ?? 0)
-  const toCryptoAmountAfterFees = bnOrZero(buyTradeAsset?.amount).minus(bnOrZero(protocolFee))
+  const sellAssetBalanceHuman = useAppSelector(state =>
+    selectPortfolioCryptoHumanBalanceByFilter(state, {
```

Comment:
non-blocking, but convention to memoize filter args if they're non-primitive types - the output of this selector is a string and won't change, but it will recompute internally

Reply:
👌 addressed in https://github.com/shapeshift/web/pull/2789/commits/26748065d9f32d351acb3c1f57ad3b717a437c22